pkg/service/keystore/storage.go

```diff
@@ -102,10 +174,59 @@ func (kss *Storage) getAndSetServiceKey(ctx context.Context) ([]byte, error) {
 		return nil, errors.Wrap(err, "could not decode service key")
 	}
 
-	kss.serviceKey = keyBytes
 	return keyBytes, nil
 }
 
+// Encryptor the interface for any encryptor implementation.
+type Encryptor interface {
+	Encrypt(ctx context.Context, plaintext, contextData []byte) ([]byte, error)
+}
+
+// Decryptor is the interface for any decryptor. May be AEAD or Hybrid.
+type Decryptor interface {
+	// Decrypt decrypts ciphertext. The second parameter may be treated as associated data for AEAD (as abstracted in
+	// https://datatracker.ietf.org/doc/html/rfc5116), or as contextInfofor HPKE (https://www.rfc-editor.org/rfc/rfc9180.html)
+	Decrypt(ctx context.Context, ciphertext, contextInfo []byte) ([]byte, error)
+}
+
+type encryptor struct {
+	db storage.ServiceStorage
+}
+
+func (e encryptor) Encrypt(ctx context.Context, plaintext, _ []byte) ([]byte, error) {
+	// get service key
+	serviceKey, err := getServiceKey(ctx, e.db)
+	if err != nil {
+		return nil, err
+	}
+	// encrypt key before storing
+	encryptedKey, err := util.XChaCha20Poly1305Encrypt(serviceKey, plaintext)
```

Comment:
can we always use `XChaCha20Poly1305`? or is this just if we generate a key ourselves?
since if a KMS is configured it could use any of these keys https://developers.google.com/tink/aead

Reply:
nvm I see this is only for our service key

for the external key - is there no service key? does this fail?

Reply:
If there is no service key, this never gets called. Details below. 

The struct `encrypter` is used only when `cfg.MasterKeyURI` is empty. It represents the old way in which we were doing things.

When using an external KMS, then the struct that's used is `wrappedEncrypter`. It delegates signing to the external KMS.